pkg/cmd/cli/sa/newtoken.go

```diff
@@ -0,0 +1,180 @@
+
+	client, err := f.Client()
+	if err != nil {
+		return err
+	}
+
+	if len(o.Namespace) == 0 {
+		namespace, _, err := f.DefaultNamespace()
+		if err != nil {
+			return fmt.Errorf("could not retrieve default namespace: %v", err)
+		}
+		o.Namespace = namespace
+	}
+
+	o.SAClient = client.ServiceAccounts(o.Namespace)
+	o.SecretsClient = client.Secrets(o.Namespace)
+	return nil
+}
+
+// Run creates a new token secret, waits for the service account token controller to fulfill it, then adds the token to the service account
+func (o *NewServiceAccountTokenOptions) Run() error {
+	serviceAccount, err := o.SAClient.Get(o.SAName)
+	if err != nil {
+		return fmt.Errorf("could not locate service account %q: %v", o.SAName, err)
+	}
+
+	tokenSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name:      secret.Strategy.GenerateName(osautil.GetTokenSecretNamePrefix(serviceAccount)),
+			Namespace: serviceAccount.Namespace,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey: serviceAccount.Name,
+				api.ServiceAccountUIDKey:  string(serviceAccount.UID),
+			},
+		},
+		Type: api.SecretTypeServiceAccountToken,
+		Data: map[string][]byte{},
+	}
+
+	persistedToken, err := o.SecretsClient.Create(tokenSecret)
+	if err != nil {
+		return err
+	}
+
+	// we need to wait for the service account token controller to make the new token valid
+	options := api.ListOptions{
+		FieldSelector:   fields.SelectorFromSet(fields.Set(map[string]string{"type": string(api.SecretTypeServiceAccountToken)})),
+		Watch:           true,
+		ResourceVersion: persistedToken.ResourceVersion,
+		TimeoutSeconds:  &o.TimeoutSeocnds,
+	}
+	watcher, err := o.SecretsClient.Watch(options)
+	if err != nil {
+		return fmt.Errorf("could not begin watch for token: %v", err)
+	}
+	defer watcher.Stop()
+
+watch:
+	for {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok || event.Type == watch.Error {
+				return fmt.Errorf("encountered error while watching for token: %v", event.Object)
+			}
+
+			if eventToken, ok := event.Object.(*api.Secret); ok {
+				if eventToken.Name == persistedToken.Name && serviceaccounts.IsValidServiceAccountToken(serviceAccount, eventToken) && event.Type == watch.Modified {
+					fmt.Fprintf(o.Out, "token %q created for service account %q\n", eventToken.Name, serviceAccount.Name)
+					break watch
+				}
+			}
+		case <-time.After(time.Duration(o.TimeoutSeocnds) * time.Second):
+			return errors.New("timed out wating for service account token controller to update token")
+		}
+	}
+
+	tokenSecret, err = o.SecretsClient.Get(persistedToken.Name)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve token %q: %v", tokenSecret.Name, err)
+	}
+	token, exists := tokenSecret.Data[api.ServiceAccountTokenKey]
+	if !exists {
+		return fmt.Errorf("service account token %q did not contain token data", tokenSecret.Name)
+	}
+
+	message := string(token)
+	if util.IsTerminalWriter(o.Out) {
+		// pretty-print for a TTY
+		message += "\n"
```

Comment:
asked @fabianofranz about this, I don't think this a good use for standard error, as there is no error.